### PR TITLE
Optimizing Gemma4 graph for prompt processing

### PR DESCRIPTION
**Pull request by Meet91721**
## Overview

I have been discussing with @am17an regarding the high difference in prompt processing rate for Gemma 4 model on llama.cpp and mlx-lm. On debugging I found that the TODO in the  `src/models/gemma4.cpp` by @ngxson states that the prompt processing can be improved by stripping the unused tokens after the last KV layer.

This PR skips the computations that are happening after the calculation of attention for the tokens that do not require to calculate the logits. The improvement in the prompt processing rate is mentioned below:

The commit: 3cf03257f219afbe7334045ff7c6a06ac68c627d  has the following prompt processing rate:
| model                          |       size |     params | backend    | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | --------------: | -------------------: |
| gemma4 E2B BF16                |   8.66 GiB |     4.65 B | MTL,BLAS   |       4 |          pp2048 |        641.61 ± 5.41 |
| gemma4 E2B BF16                |   8.66 GiB |     4.65 B | MTL,BLAS   |       4 |           tg128 |         19.31 ± 0.18 |

Current PR's prompt processing rate:

| model                          |       size |     params | backend    | threads |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | --------------: | -------------------: |
| gemma4 E2B BF16                |   8.66 GiB |     4.65 B | MTL,BLAS   |       4 |          pp2048 |       1126.27 ± 1.38 |
| gemma4 E2B BF16                |   8.66 GiB |     4.65 B | MTL,BLAS   |       4 |           tg128 |         19.61 ± 0.50 |

I have tested this using the following command:
```
./llama-bench -m ../models/gguf/gemma-4-E2B-it-BF16.gguf \
    -p 2048 \
    -n 128 \
    -r 5
```

## Additional information
I am testing this on the MacBook Air M3, 16GB RAM.
I have used the following model: [ggml-org/gemma-4-E2B-it-GGUF](https://huggingface.co/ggml-org/gemma-4-E2B-it-GGUF/blob/main/gemma-4-E2B-it-BF16.gguf)

There are still a few things that I want to improve in this PR:

- Currently all the tokens in the microbatch should not require to calculate logits.
Need to make changes here by optimizing for the microbatches which contains both the tokens: ones which require logits and other which do not.
- There are a few more operations that are performed after the KV cache is calculated, need to safely optimize them.
I tried putting the same condition after the layers loop in the graph construction, however it throws error. Need to debug that.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES, I have used GPT 5.6 SOL for understanding some parts of the codebase. All the changes made are done manually.

### CODE REVIEW

File: src/models/gemma4.cpp [hunk 1]
```diff
@@ -273,6 +273,10 @@ llama_model_gemma4::graph::graph(const llama_model & model, const llm_graph_para
                     Qcur, nullptr, nullptr, nullptr, nullptr, nullptr, hparams.f_attention_scale, il);
         }
 
+        if(!(hparams.has_kv(il)) && (*inp_out_ids).ne[0] == 0) {
+            continue;
+        }
+
         // TODO @ngxson : strip unused token right after the last KV layer to speed up prompt processing
         // keep all rows when extracting unmasked nextn embeddings (MTP target needs the hidden state for every token)
```
Comment (ngxson) [inline src/models/gemma4.cpp:281]: this is an incorrect fix, you did not see my next TODO
